Remove Webpack from the Metalsmith build script

cvalarida opened the pull request:
## Description
This is the first step in separating the content and application deploys. The goal here is to isolate each build without impacting engineers (too much) and the current deployment process.

The [connected ticket](https://github.com/department-of-veterans-affairs/va.gov-team/issues/7332) is fairly comprehensive in what this PR does. I've also made comments in a PR review to explain what I'm dong.

It's important to note that the **Webpack build is standalone** but the **content build is not** right now. This is because the Webpack build currently still adds hashes to the asset filenames and the Metalsmith script needs to translate the `[data-entry-name]`s found in the DOM into `src` and `href` attributes that point to the appropriate files. Once we stop adding hashes to the filenames, we should be able remove this `processEntryNames` plugin, update the Liquid templates to no longer use `[data-entry-name]`, and completely isolate the content build script.

Currently, to run the content build, the webpack build must be run first. This can be accomplished with a simple `yarn build`.

## Testing done
So much manual testing.

### How to test this PR
There are four main touch points to this:
1. `yarn build:webpack`
1. `yarn watch`
1. `yarn build:content`
1. `yarn build`

Try experimenting with different options on each of those commands.

## Screenshots
**Header + application**
![image](https://user-images.githubusercontent.com/12970166/78727238-f49e9200-78e8-11ea-96a8-5251b203c623.png)
Note: The mega menu is currently sourced from Drupal, so it's wonky without that data. Also, the images aren't processed by Webpack, so the standalone Webpack build doesn't have them. We can change that in a separate PR.

**Footer**
![image](https://user-images.githubusercontent.com/12970166/78727338-38919700-78e9-11ea-9bdb-20077b525b29.png)
Note: The footer data was copied from staging.va.gov and created dynamically somewhere. We could probably change that up for these local development pages, but I was trying to reduce the footprint of this PR.

## Acceptance criteria
- [x] Local application development can be accomplished using only the Webpack build
- [x] The current deployment process isn't messed up

## Definition of done
- [ ] Events are logged appropriately
- [ ] Documentation has been updated, if applicable
- [ ] A link has been provided to the originating GitHub issue (or connected to it via ZenHub)
- [ ] No sensitive information (i.e. PII/credentials/internal URLs/etc.) is captured in logging, hardcoded, or specs

Commit:
Rename to build:webpack for consistency

Commit:
Rename to build:content

Commit:
Consolidated build script

Commit:
Change watch command to run the dev server

Commit:
Rename build.js -> build-content.js

Commit:
Start a consolidated build script

Commit:
Only download webpack assets from Metalsmith

This is because it would take more work than it's possibly worth to make
the downloadAssets script standalone.

Commit:
Update the echo in the build script

Commit:
Run the yarn commands

Commit:
Remove vestigial metalsmith-webpack code

Commit:
Move manifest-helpers.js

Commit:
Update standalone webpack config

Commit:
Delete webpack config

Commit:
Rename standalone-webpack.config.js -> webpack.config.js

Commit:
Read existing Webpack assets into Metalsmith

Commit:
Fix prebuild script

Commit:
Refactor getAppManifests

Commit:
Preserve Webpack output in the content build

Commit:
Fix readAssetsFromDisk plugin

Commit:
Remove comment

Commit:
Add landing pages

Commit:
Only build landing pages during watch task

Commit:
Add comment for why writeToDisk needs to be true

Commit:
Don't build new landing pages if old ones exist

Commit:
Use path.join to find the right path

Commit:
Only try to preserve the webpack assets if they exist

Commit:
Double-logging looked dumb

Commit:
Only read Webpack assets if they're there

Commit:
Create a src/platform/landing-pages/ and move the template

Commit:
Move headerFooterData into dedicated JSON file

cvalarida commented on config/webpack.config.js:
```diff
@@ -33,7 +44,80 @@ const globalEntryFiles = {
+ * If the `entry` CLI argument is passed, only the specified
+ * application entries are built.
+ */
+function getEntryPoints(entry) {
```
I had these in the experimental `standalone-webpack.config.js` which I renamed to `webpack.config.js`. Essentially, this lays the groundwork for working with multiple `manifest.json` files.

rianfowler replied:
> lays the groundwork for working with multiple manifest.json files.

The build process already works with multiple manifest.json files. How his this different?

cvalarida replied:
It already works with multiple manifest.json files because of these "manifest helpers." I just copied them into this file because we'll be removing `src/site/` down the road and I didn't want any dependencies in there.

cvalarida replied:
Ah, I remember now. I copied this from [`build/webpack/index.js`](https://github.com/department-of-veterans-affairs/vets-website/blob/7ff9874b2da7eef36997a9033ee96c043c5a915f/src/site/stages/build/webpack/index.js#L18). :sweat_smile: 

I still didn't author it. Just misremembered where I copied it from.

cvalarida commented on config/webpack.config.js:
```diff
@@ -189,6 +275,7 @@ const configGenerator = (buildOptions, apps) => {
       }),
       new webpack.IgnorePlugin(/^\.\/locale$/, /moment$/),
     ],
+    devServer: generateWebpackDevConfig(buildOptions),
```
Previously, we'd combine the two configs in the Metalsmith plugin. This is more semantic.

cvalarida commented on config/webpack.config.js:
```diff
@@ -197,6 +284,50 @@ const configGenerator = (buildOptions, apps) => {
         fileName: 'file-manifest.json',
       }),
     );
+  } else {
```
This bit is new.

cvalarida commented on config/webpack.dev.config.js:
```diff
@@ -45,6 +45,8 @@ function generateWebpackDevConfig(buildOptions) {
       modules: false,
       warnings: true,
     },
+    // Needed to write the landing pages to disk so webpack-dev-server will actually serve them
+    writeToDisk: true,
```
I'm actually not sure why it was working before, if I'm honest. Something something Metalsmith... :man_shrugging: 

If anybody knows how to _not_ write everything to disk during a `watch` task but still get `webpack-dev-server` to actually serve the pages created by `html-webpack-plugin`, I'm all ears! I hope this doesn't affect hot module reloading or any other fancy Webpack whatnots we may want to try out down the road.

cvalarida commented on package.json:
```diff
@@ -5,10 +5,11 @@
+    "build:content": "node --max-old-space-size=4096 script/build-content.js",
     "build:content:test": "node script/test-cms-export.js",
-    "build-webpack": "NODE_OPTIONS=--max-old-space-size=4096 webpack --config config/standalone-webpack.config.js",
+    "build:webpack": "NODE_OPTIONS=--max-old-space-size=4096 webpack --config config/webpack.config.js --progress",
```
Decided to rename this for consistency. Note that it's also now using the `webpack.config.js` instead of the former `standalone-webpack.config.js` now. This is because the `webpack.config.js` is _meant_ to be standalone now. (As opposed to being used by an instance of Webpack run through Metalsmith.)

cvalarida commented on script/build.sh:
```diff
@@ -0,0 +1,42 @@
+# Default to local
+assetSource="local"
+
+# Save the arguments to this script for later; they get drained in the following for loop
+args="$*"
```
I think I saw a better way to do this the other day, but didn't write it down and goodness knows I won't be able to find it again. Stinking bashisms.

cvalarida commented on src/platform/landing-pages/dev-template.ejs:
```diff
@@ -0,0 +1,168 @@
+<html>
```
This whole template is essentially [`src/site/layouts/page-react.html`](https://github.com/department-of-veterans-affairs/vets-website/blob/master/src/site/layouts/page-react.html) with a few things stripped out since it will only be used for local development. (Goodbye, IE-specific ~garbage~ comment directive thingies!)

cvalarida commented on src/platform/landing-pages/dev-template.ejs:
```diff
@@ -0,0 +1,168 @@
+
+    <script type="text/javascript">
+      window.VetsGov = window.VetsGov || {};
+      window.VetsGov.headerFooter = <%= JSON.stringify(headerFooterData) %>;
```
Note that the templating engine is different. This isn't Liquid anymore. It's _basically_ EJS, but _actually_ lodash templates. See the `html-webpack-plugin`'s ["template option" doc](https://github.com/jantimon/html-webpack-plugin/blob/master/docs/template-option.md) for more details.

cvalarida commented on src/platform/landing-pages/header-footer-data.json:
```diff
@@ -0,0 +1,321 @@
+{
```
This is copied from `staging.va.gov` just so it wouldn't die on me when the page loaded up. Probably we want to not have this for local application development, but that change can come in another PR.

cvalarida commented on src/site/stages/build/index.js:
```diff
@@ -39,6 +46,40 @@ const rewriteVaDomains = require('./plugins/rewrite-va-domains');
+ * This can be removed when we move the content build to a new repository and
+ * this script no longer interacts with the Webpack output at all.
+ */
+function preserveWebpackOutput(metalsmithDestination) {
```
If the comment for this function doesn't make sense, please let me know how I can improve it.

cvalarida commented on src/site/stages/build/plugins/create-react-pages.js:
```diff
@@ -1,12 +1,11 @@
-    const root = path.join(__dirname, '../../../../..');
     manifestHelpers
-      .getAppManifests(root)
+      .getAppManifests()
```
The result of a little refactor I made. It's intentional.

cvalarida reviewed (commented):
It's a significant change, but not as daunting as the git diff makes it look.

bkjohnson commented on src/site/stages/build/plugins/read-assets-from-disk.js:
```diff
@@ -0,0 +1,29 @@
+  return files => {
+    if (fs.existsSync(webpackOutput)) {
+      fs.readdirSync(webpackOutput).forEach(
+        // eslint-disable-next-line no-return-assign
```
The reason the bot hasn't jumped all over this is because no commits have been pushed since this became a PR.

bkjohnson commented on config/webpack.config.js:
```diff
@@ -33,7 +44,80 @@ const globalEntryFiles = {
+ * application entries are built.
+ */
+function getEntryPoints(entry) {
+  const manifests = getAppManifests(path.join(__dirname, '../'));
```
Due to your refactor do we need to pass it an arg now?

cvalarida replied:
Oh, good call. :male_detective:

bkjohnson commented on config/webpack.config.js:
```diff
@@ -33,7 +44,80 @@ const globalEntryFiles = {
+ */
+function getSettings() {
+  const settings = {};
+  const manifests = getAppManifests(path.join(__dirname, '../'));
```
Same thing here.

bkjohnson commented on config/webpack.config.js:
```diff
@@ -33,7 +44,80 @@ const globalEntryFiles = {
+
+module.exports = env => {
+  const buildOptions = Object.assign(
+    {},
```
What happens if we don't have this empty object as the target?

cvalarida replied:
It'll modify the next object literal. Which is fine. That's a good point. :+1:

bkjohnson commented on config/webpack.config.js:
```diff
@@ -197,6 +284,50 @@ const configGenerator = (buildOptions, apps) => {
+            inject: false,
+          });
+        })
+        .filter(p => p),
```
What is this filter doing exactly?

cvalarida replied:
When a manifest file either doesn't have a `rootUrl` specified _or_ there's already a landing page at the `rootUrl` (say, from a content build), the `.map()` will return `undefined`. I didn't want to add `undefined` items to the `baseConfig.plugins` array, so this filter just keeps everything that's truthy.

rianfowler replied:
Should filter before mapping here, too.

bkjohnson commented on src/site/stages/build/index.js:
```diff
@@ -175,7 +221,17 @@ function defaultBuild(BUILD_OPTIONS) {
   smith.use(downloadDrupalAssets(BUILD_OPTIONS), 'Download Drupal assets');
 
-  configureAssets(smith, BUILD_OPTIONS);
+  if (BUILD_OPTIONS['asset-source'] !== assetSources.LOCAL) {
```
Does this need to be added to `src/site/stages/build/options.js`?

cvalarida replied:
It's [already there](https://github.com/department-of-veterans-affairs/vets-website/tree/master/src/site/stages/build/options.js#L31).

bkjohnson reviewed (commented):
I haven't pulled this branch and played around with it - I'll try to do that soon to get a better idea of what is happening.

Commit:
Remove vestigial parameter to getAppManifests()

Commit:
Remove pointless empty object literal

Commit:
Import the manifest helpers from the app coverage script

cvalarida commented on src/site/stages/build/index.js:
```diff
@@ -39,6 +46,40 @@ const rewriteVaDomains = require('./plugins/rewrite-va-domains');
+  if (webpackDirExists) fs.moveSync(webpackDir, tempDir);
+
+  return () => {
+    /* eslint-disable no-console */
```
This is intentional. I wanted to output this for debugging purposes if needed later.

cvalarida commented on src/site/stages/build/plugins/read-assets-from-disk.js:
```diff
@@ -0,0 +1,29 @@
+  return files => {
+    if (fs.existsSync(webpackOutput)) {
+      fs.readdirSync(webpackOutput).forEach(
+        // eslint-disable-next-line no-return-assign
```
Open to suggestions on how to do this chunk differently.

cvalarida commented on src/site/stages/build/plugins/read-assets-from-disk.js:
```diff
@@ -0,0 +1,29 @@
+      fs.readdirSync(webpackOutput).forEach(
+        // eslint-disable-next-line no-return-assign
+        filename =>
+          // eslint-disable-next-line no-param-reassign
```
Same here.

cvalarida commented on src/site/stages/build/plugins/read-assets-from-disk.js:
```diff
@@ -0,0 +1,29 @@
+          }),
+      );
+    } else {
+      // eslint-disable-next-line no-console
```
Also intentional for debugging purposes.

ncksllvn replied:
TBH, this could be a great opportunity to use the `debug` module and really add some logging into this, since it is complex

cvalarida replied:
Oh, interesting. I'll check it out. :+1

Commit:
Use build:webpack in circle

Commit:
Overwrite any temporary files that might already be there

Commit:
Missed a spot

Commit:
Merge branch 'master' into remove-webpack-from-metalsmith-cv

Commit:
Testing console logs to debug Jenkins error

cvalarida commented on src/site/stages/build/index.js:
```diff
@@ -38,6 +45,50 @@ const rewriteVaDomains = require('./plugins/rewrite-va-domains');
+
+  // Immediately move the Webpack output to a new directory
+  if (webpackDirExists) {
+    /* eslint-disable no-console */
```
Verifying my sanity with a few console logs. Jenkins is getting a very confusing error where it dies because the directory I'm trying to copy doesn't exist...even though I'm only copying it if it does exist....so...that's fun...

Commit:
Remove build directory based on buildtype

Commit:
Don't quite args in the build script

This was causing `yarn build` to tack on a `''` at the end of these
invocations, which made Webpack angry.

Commit:
Use the build script for `yarn build`

bkjohnson reviewed (commented):
Nothing big, but I'm not sure I see the full potential of the `build.sh` script. Is it doing things that `yarn build` can't do with the flags?

Commit:
Pass CLI args to webpack

cvalarida commented on script/build.sh:
```diff
@@ -0,0 +1,42 @@
+#!/bin/sh
+
+##
+# Run the application and content builds. This script builds or fetches the
```
Right, so this file is here so I don't have to change any of the interface with CI yet. We'll get there, but I wanted to roll that out separately since I think it'll be more involved.

cvalarida replied:
I tried to make the `npm` script do a simple `yarn build:webpack && yarn build:content`, but needed to pass the CLI args from the `yarn build` command to both, so that didn't work.

bkjohnson replied:
Ah, I made this comment before you refactored a few things.  It makes more sense now!

cvalarida commented on config/webpack.config.js:
```diff
@@ -33,7 +44,79 @@ const globalEntryFiles = {
+}
+
+module.exports = env => {
+  const buildOptions = Object.assign(
+    {
+      api: '',
+      buildtype: 'localhost',
+      host: 'localhost',
+      port: 3001,
+      watch: false,
+    },
+    env,
+  );
+  // Assign additional defaults which reference other properties
+  Object.assign(buildOptions, {
+    destination: path.resolve(
+      __dirname,
+      '../',
+      'build',
+      buildOptions.buildtype,
+    ),
+  });
+  buildOptions.settings = getSettings();
```
Oh interesting. I'll check out the getter. :+1: If that solves my problem, I think combining all this would be great.

cvalarida commented on config/webpack.config.js:
```diff
@@ -33,7 +44,79 @@ const globalEntryFiles = {
+ */
+function getEntryPoints(entry) {
+  const manifests = getAppManifests();
+  let manifestsToBuild = manifests;
+  if (entry) {
+    const entryNames = entry.split(',').map(name => name.trim());
+    manifestsToBuild = manifests.filter(manifest =>
+      entryNames.includes(manifest.entryName),
+    );
+  }
```
:thinking: True. It would produce unexpected results if we have two apps where one name is a subset of another, though. For example, `{ appName: '526' }` and `{ appName: '526-all-claims' }`. We don't right now, but it's a thing to watch out for.

bkjohnson replied:
Good point.

Clark-W-Griswold replied:
Should we specifically look for that and throw a build error with solid context to prevent a developer accidentally adding a duplicate name?

cvalarida replied:
We definitely could provide better error handling in here! Since we don't do that right now, I'm going to leave it off of this PR. The goal here is to provide 1:1 functionality with the old way before we make any improvements. (Trying to make as small of changes as I can get away with.)

Clark-W-Griswold replied:
👍

Commit:
Simplify buildOptions

cvalarida commented on script/build.sh:
```diff
@@ -0,0 +1,61 @@
+# Save the arguments to this script for later; they get drained in the following for loop
+args="$*"
+
+# Prepend env. to all args so they get passed to the Webpack config
+webpackArgs=$(echo -n "${args}" | sed -E 's/--([^ =]+)/--env.\1/g')
```
Because nothing can be easy, right? In order for CLI args to get passed to `webpack`, they need to be `--env.whatever`. Documentation can be found [here](https://webpack.js.org/guides/environment-variables/).

rianfowler replied:
What's the reason for putting some logic in bash and some logic in node? Node has access to the command line arguments. You could use something like [`command-line-args`](https://github.com/75lb/command-line-args#readme) to define accepted arguments.

cvalarida replied:
We potentially could make this a `.js` script, but since it's in charge of orchestrating two separate yarn scripts, I decided to keep it in the shell. For what it's worth, this is only a temporary thing until we can remove the content build from `vets-website` entirely.

ncksllvn commented on src/platform/landing-pages/header-footer-data.json:
```diff
@@ -0,0 +1,321 @@
+      "target": null,
+      "title": "Veterans Portrait Project"
+    }
+  ],
```
The footer links are just from https://github.com/department-of-veterans-affairs/vets-website/blob/master/src/platform/static-data/footer-links.json but exposed in `drupalHeaderFooter.json` for TeamSite pages to download

ncksllvn commented on src/platform/landing-pages/header-footer-data.json:
```diff
@@ -0,0 +1,321 @@
+      ]
+    }
+  ]
+}
```
This is all just from Drupal

cvalarida replied:
Yeah, the idea here is that it's just a placeholder on the application landing pages for local development. There's something in `startApp` (or something) that required this data, so rather than refactor it, I figured I'd just feed it this placeholder for now. The next step is to move all that out, I think, but I don't know what that means yet.

ncksllvn commented on src/site/stages/build/index.js:
```diff
@@ -38,6 +45,50 @@ const rewriteVaDomains = require('./plugins/rewrite-va-domains');
+ */
+function preserveWebpackOutput(metalsmithDestination) {
+  const webpackBuildDirName = 'generated';
+  const tempDir = path.join(__dirname, '../../../../tmp/', webpackBuildDirName);
```
Where does this `tmp` directory come from?

cvalarida replied:
Right here. :smile: 

Like I explained in the comment for this function, `smith.build()` nukes _everything_ in the `build/${buildtype}/`, so this function
1. Moves the output from the Webpack build into a temporary folder outside of `build/` so we don't lose it
1. Returns a plugin that moves it back during the execution of the pipeline, after `.build()` clears the directory out

ncksllvn commented on src/site/stages/build/index.js:
```diff
@@ -38,6 +45,50 @@ const rewriteVaDomains = require('./plugins/rewrite-va-domains');
+    /* eslint-enable no-console */
+  };
+}
+
```
When does this execute? Shouldn't it execute at the end of each Webpack build or watch task?

ncksllvn replied:
Or nvm about the watch task, because I suppose that isn't necessary if to develop for content you need to do a build beforehand

cvalarida replied:
Right, so like I explained in the above comment, this is meant to preserve any Webpack output. If there _is_ no Webpack output, then it'll essentially be a no-op.

It's needed only when we run a content build after Webpack does its thing. When we run Webpack by itself with either `yarn build:webpack` or `yarn watch`, we won't need this, like you said.

I'm not doing this at the end of the Webpack build because I'm trying to keep the same output... :thinking: It makes sense in my head, but communicating it is hard...Essentially, I've thought about moving the Webpack output to some directory completely outside `build/`, but we run into other things that need to be changed for that to work seamlessly, and I felt like that was beyond the scope of this particular PR. That'll be coming in a future iteration (if at all).

ncksllvn commented on config/webpack.config.js:
```diff
@@ -197,6 +275,50 @@ const configGenerator = (buildOptions, apps) => {
+      getAppManifests()
+        .map(m => {
+          // There's no place to create a landing page at
+          if (!m.rootUrl) return undefined;
```
Should this throw an error? I didn't know we had manifests without this.

cvalarida replied:
I didn't either. :man_shrugging: But it turns out it's a thing. Rather than figure out why and change folks' applications, I figured I'd just not generate a dev landing page for them unless a `rootUrl` exists.

I don't think it makes sense to break everybody's watch task if some unrelated application's `manifest.json` doesn't specify a `rootUrl`.

ncksllvn replied:
Actually, `static-pages` doesn't have a `rootUrl`, so that makes sense 😅 Although that bundle is a special case

ncksllvn commented on src/site/stages/build/plugins/read-assets-from-disk.js:
```diff
@@ -0,0 +1,29 @@
+      console.log('No Webpack assets found');
+    }
+  };
+}
```
So this is like the opposite side of `preserveWebpackOutput`?

cvalarida replied:
Not exactly. So previously, when we were running Webpack through Metalsmith, the plugin would take the output and put each file from Webpack into Metalsmith's `files[]`. We then used that for broken link checking and doing the `processDataEntry` thing. Rather than changing that part yet, I decided to just read the Webpack output into `files` like we were doing before.

This _will_ change, but in a future PR.

It might help to think of this as isolating the application build from the content build, but not the content build from the application build. The content build still has a dependency on the application build. The next step in this isolation process is to remove that dependency.

ncksllvn reviewed (commented):
Can you tell me the process/commands if I were to be editing the `static-pages` bundle that requires running the whole site? Potentially we might also need to edit templates at the same time

cvalarida commented:
> Can you tell me the process/commands if I were to be editing the static-pages bundle that requires running the whole site?

Great question! I was hoping you could help me out with this, actually, since I'm really unfamiliar with the `static-pages` bundle.

I think the process would go something like this:
1. `yarn build`
    - This will build both applications and content (in that order)
    - The HTML files will reference JS assets without the hashes in their names (this is important)
1. `yarn watch`
    - This will build only applications, but all the static content will still be there
    - Development application landing pages are only created if they don't already exist in the `build/${buildtype}/path/to/page/index.html`
      - So when you do a `yarn build` first, those pages aren't overwritten
    - The watch task now writes files to disk, so when we load up the HTML pages and they request files on disk, they'll get the most up-to-date version
      - This is why it's important that the hashes in the filenames aren't a thing
      - Production builds may not work with this process until we remove those hashes (which can be done at any time now that we're using `Cache-Control: no-cache`)

The main thing to keep in mind is that the content build will wipe out the application assets, but the application build will preserve the static content files.

When you get a chance, try it out and let me know how it works for you! Like I said, this is maybe the part of development on VA.gov I'm least familiar with, so any input you have is welcome!

Clark-W-Griswold commented on config/manifest-helpers.js:
```diff
@@ -2,7 +2,9 @@
 const path = require('path');
 
-function getAppManifests(root) {
+const root = path.join(__dirname, '..');
```
👍

rianfowler commented on config/webpack.config.js:
```diff
@@ -1,13 +1,24 @@
 // Staging config. Also the default config that prod and dev are based off of.
+const path = require('path');
+const fs = require('fs');
+
```
```suggestion
```

rianfowler commented on config/webpack.config.js:
```diff
@@ -1,13 +1,24 @@
+const fs = require('fs');
+
+const webpack = require('webpack');
+
```
```suggestion
```

rianfowler commented on config/webpack.config.js:
```diff
@@ -33,7 +44,71 @@ const globalEntryFiles = {
+/**
+ * Get a list of all the entry points.
+ *
+ * If the `entry` CLI argument is passed, only the specified
+ * application entries are built.
+ */
+function getEntryPoints(entry) {
```
```suggestion
 * @param {string} [entries] - List of comma delimited entries to build. Builds all entries if no value passed. 
 */
 function getEntryPoints(entries) { 
```
if we're going to document it, might as well use jsdoc syntax. what is the returned value?

cvalarida replied:
Good call. I didn't write these (just copied), but I'll add some jsdocs. :+1:

rianfowler commented on config/webpack.config.js:
```diff
@@ -33,7 +44,71 @@ const globalEntryFiles = {
+function getEntryPoints(entry) {
+  const manifests = getAppManifests();
+  let manifestsToBuild = manifests;
+  if (entry) {
+    const entryNames = entry.split(',').map(name => name.trim());
+    manifestsToBuild = manifests.filter(manifest =>
+      entryNames.includes(manifest.entryName),
+    );
+  }
+
+  return getWebpackEntryPoints(manifestsToBuild);
```
```suggestion
  if (!entry) {
    return getWebpackEntryPoints(manifests);
  }
  const entryNames = entry.split(',').map(name => name.trim());
  return getWebpackEntryPoints(manifests.filter(manifest =>
    entryNames.includes(manifest.entryName),
  ));
```
I think early returns can be easier to read.

rianfowler commented on config/webpack.config.js:
```diff
@@ -33,7 +44,71 @@ const globalEntryFiles = {
+ *
+ * @return {Object} settings
+ */
+function getSettings() {
```
```suggestion
function getSettingsForDevServer() {
```
or `ForWatchTask` or `getLocalBuildSettings`

renaming can remove the need for the long comment. Description could be 

> Loads redirect settings for local watch task

rianfowler commented on config/webpack.config.js:
```diff
@@ -33,7 +44,71 @@ const globalEntryFiles = {
+  settings.applications = manifests
+    .map(
+      m =>
+        // Some manifests don't have a rootUrl
```
What does this mean? 

> Manifests without rootUrls are excluded from the redirect settings

rianfowler commented on config/webpack.config.js:
```diff
@@ -33,7 +44,71 @@ const globalEntryFiles = {
+  const manifests = getAppManifests();
+  settings.applications = manifests
+    .map(
+      m =>
```
```suggestion
      manifest =>
```

rianfowler commented on config/webpack.config.js:
```diff
@@ -33,7 +44,71 @@ const globalEntryFiles = {
+        },
+    )
+    // Filter out empty entries
+    .filter(a => !!a);
```
Could we filter first and then map? I think this might remove the need for the ternary.

rianfowler commented on config/webpack.config.js:
```diff
@@ -33,7 +44,71 @@ const globalEntryFiles = {
+    watch: false,
+    ...env,
+    get destination() {
+      return path.resolve(__dirname, '../', 'build', this.buildtype);
```
maybe add a comment explaining why this has to be determined at runtime.

rianfowler commented on config/webpack.config.js:
```diff
@@ -33,7 +44,71 @@ const globalEntryFiles = {
+  return settings;
+}
+
+module.exports = env => {
```
Maybe put this into a named function and export the named function.

cvalarida replied:
:man_shrugging: This seems to be a [pretty normal convention](https://webpack.js.org/configuration/configuration-types/).

rianfowler replied:
gross

rianfowler reviewed (approved):
We should add unit tests to some of this stuff.

cvalarida commented:
I dig the idea of unit testing this! But, um...how?

Commit:
Clean imports a bit

Commit:
Use JSDoc for getEntryPoints

Commit:
Refactor getSettings

Commit:
Rename to getDevServerSettings

Commit:
Update comment

Commit:
Refactor HtmlWebpackPlugin section

Commit:
Update JSDoc to reflect what getEntryPoints actually returns

Commit:
Watch static content and pages

Commit:
Use the buildtype in the temporary storage of webpack assets

Commit:
Clean up console logs

Commit:
Clean up tmp/

Commit:
Don't lie, now!

Commit:
Don't use the concurrently shorthand

Turns out this wasn't passing the args as expected. I think.

Commit:
Name the processes

cvalarida commented on src/site/stages/build/index.js:
```diff
@@ -38,6 +45,55 @@ const rewriteVaDomains = require('./plugins/rewrite-va-domains');
+
+  // Immediately move the Webpack output to a new directory
+  if (webpackDirExists) {
+    // eslint-disable-next-line no-console
```
Changed these from the `/* eslint-disable no-console */` that they were before since I'm now only logging one line.
